In [24]:
import pandas as pd

# Read the 2012 dataset with a different encoding
data_2012 = pd.read_csv('/Users/nafis/Downloads/DATASET/2012_Data.csv', encoding='ISO-8859-1')

# Read the 2013 dataset with a different encoding
data_2013 = pd.read_csv('/Users/nafis/Downloads/DATASET/2013_Data.csv', encoding='ISO-8859-1')

# Display the first few rows of each dataset to verify that they are loaded correctly
print("2012 Data:")
print(data_2012.head())

print("\n2013 Data:")
print(data_2013.head())



/var/folders/zy/70yrs2wx10bdf9kdl2h6mq240000gn/T/ipykernel_11504/51244734.py:4: DtypeWarning: Columns (7,9) have mixed types. Specify dtype option on import or set low_memory=False.
  data_2012 = pd.read_csv('/Users/nafis/Downloads/DATASET/2012_Data.csv', encoding='ISO-8859-1')
/var/folders/zy/70yrs2wx10bdf9kdl2h6mq240000gn/T/ipykernel_11504/51244734.py:7: DtypeWarning: Columns (38) have mixed types. Specify dtype option on import or set low_memory=False.
  data_2013 = pd.read_csv('/Users/nafis/Downloads/DATASET/2013_Data.csv', encoding='ISO-8859-1')


2012 Data:
   accounting_date  fiscal_year  fiscal_month  calendar_year  calendar_month  \
0         20120509         2012            11           2012               5   
1         20120216         2012             8           2012               2   
2         20120509         2012            11           2012               5   
3         20120518         2012            11           2012               5   
4         20120109         2012             7           2012               1   

   calendar_day  company_code customer_code  customer_district_code  \
0             9           101     411800601                     410   
1            16           101     361000403                     300   
2             9           101     361000403                     300   
3            18           101     565540415                     500   
4             9           101     565540415                     500   

                        item_code  ... value_quantity value_price_adjustment  \
0

In [25]:
# Display the column names for the 2012 dataset
print("Columns in 2012 Data:")
print(data_2012.columns)

# Display the column names for the 2013 dataset
print("\nColumns in 2013 Data:")
print(data_2013.columns)


Columns in 2012 Data:
Index(['accounting_date', 'fiscal_year', 'fiscal_month', 'calendar_year',
       'calendar_month', 'calendar_day', 'company_code', 'customer_code',
       'customer_district_code', 'item_code', 'business_area_code',
       'item_group_code', 'item_class_code', 'item_type', 'bonus_group_code',
       'environment_group_code', 'technology_group_code',
       'commission_group_code', 'reporting_classification', 'light_source',
       'warehouse_code', 'abc_class_code', 'abc_class_volume',
       'business_chain_l1_code', 'business_chain_l1_name',
       'contact_method_code', 'salesperson_code', 'order_type_code',
       'market_segment', 'value_sales', 'value_cost', 'value_quantity',
       'value_price_adjustment', 'currency', 'item_source_class',
       'invoice_number', 'line_number', 'invoice_date',
       'customer_order_number', 'order_date', 'dss_update_time'],
      dtype='object')

Columns in 2013 Data:
Index(['accounting_date', 'fiscal_year', 'fiscal_month

In [26]:
# Check if the column names are the same
columns_match = data_2012.columns.equals(data_2013.columns)

print("Do both datasets have the same columns?")
print(columns_match)

# Alternatively, to see any differences in column names:
columns_2012 = set(data_2012.columns)
columns_2013 = set(data_2013.columns)

# Columns that are in 2012 but not in 2013
print("\nColumns in 2012 but not in 2013:")
print(columns_2012 - columns_2013)

# Columns that are in 2013 but not in 2012
print("\nColumns in 2013 but not in 2012:")
print(columns_2013 - columns_2012)

Do both datasets have the same columns?
True

Columns in 2012 but not in 2013:
set()

Columns in 2013 but not in 2012:
set()


In [27]:
# List of columns to analyze
columns_to_check = [
    'company_code', 'customer_code', 'customer_order_number', 
    'invoice_number', 'item_code', 'customer_district_code', 
    'warehouse_code', 'currency', 'contact_method_code', 
    'line_number', 'invoice_date'
]

# Get the number of non-null data points for each column in 2012 data
data_points_2012 = data_2012[columns_to_check].notnull().sum()

# Get the number of non-null data points for each column in 2013 data
data_points_2013 = data_2013[columns_to_check].notnull().sum()

# Display the results
print("Number of data points in 2012 Data:")
print(data_points_2012)

print("\nNumber of data points in 2013 Data:")
print(data_points_2013)

Number of data points in 2012 Data:
company_code              1037205
customer_code             1037205
customer_order_number     1037205
invoice_number            1037205
item_code                 1037205
customer_district_code    1037205
warehouse_code            1037205
currency                  1037205
contact_method_code       1037205
line_number               1037205
invoice_date              1037205
dtype: int64

Number of data points in 2013 Data:
company_code              951177
customer_code             951177
customer_order_number     951177
invoice_number            951177
item_code                 951177
customer_district_code    951177
warehouse_code            951177
currency                  951177
contact_method_code       951177
line_number               951177
invoice_date              951177
dtype: int64


## Checking the number of null values

In [28]:
# List of columns to analyze
columns_to_check = [
    'company_code', 'customer_code', 'customer_order_number', 
    'invoice_number', 'item_code', 'customer_district_code', 
    'warehouse_code', 'currency', 'contact_method_code', 
    'line_number', 'invoice_date'
]

# Get the number of null values for each column in 2012 data
null_values_2012 = data_2012[columns_to_check].isnull().sum()

# Get the number of null values for each column in 2013 data
null_values_2013 = data_2013[columns_to_check].isnull().sum()

# Display the results
print("Number of null values in 2012 Data:")
print(null_values_2012)

print("\nNumber of null values in 2013 Data:")
print(null_values_2013)

Number of null values in 2012 Data:
company_code              0
customer_code             0
customer_order_number     0
invoice_number            0
item_code                 0
customer_district_code    0
warehouse_code            0
currency                  0
contact_method_code       0
line_number               0
invoice_date              0
dtype: int64

Number of null values in 2013 Data:
company_code              0
customer_code             0
customer_order_number     0
invoice_number            0
item_code                 0
customer_district_code    0
warehouse_code            0
currency                  0
contact_method_code       0
line_number               0
invoice_date              0
dtype: int64


## Checking the number of unique values

In [29]:
# Unique values count
unique_counts_2012 = data_2012[columns_to_check].nunique()
unique_counts_2013 = data_2013[columns_to_check].nunique()

print("Unique value counts in 2012 Data:")
print(unique_counts_2012)

print("\nUnique value counts in 2013 Data:")
print(unique_counts_2013)

Unique value counts in 2012 Data:
company_code                   9
customer_code               6576
customer_order_number     339575
invoice_number            341877
item_code                  27322
customer_district_code        16
warehouse_code                55
currency                       6
contact_method_code          788
line_number                  125
invoice_date                 274
dtype: int64

Unique value counts in 2013 Data:
company_code                   6
customer_code               3915
customer_order_number     321736
invoice_number            333798
item_code                  23281
customer_district_code        17
warehouse_code                29
currency                       5
contact_method_code         1274
line_number                  138
invoice_date                 269
dtype: int64


### Columns with Notable Unique Values:

1. **customer_code**
   - **Unique Values**: 6,576 (2012) vs. 3,915 (2013)
   - **Importance**: This column represents individual customers. A higher number of unique customer codes suggests a diverse customer base, which can be important for customer segmentation, targeted marketing, and understanding purchasing behaviors.

2. **customer_order_number**
   - **Unique Values**: 339,575 (2012) vs. 321,736 (2013)
   - **Importance**: This indicates the number of unique orders placed. It’s critical for understanding order frequency, sales trends, and operational efficiency. A large number of orders can imply high demand or frequent purchasing, which is valuable for inventory management.

3. **invoice_number**
   - **Unique Values**: 341,877 (2012) vs. 333,798 (2013)
   - **Importance**: Each invoice represents a transaction. Analyzing invoices helps in revenue tracking, sales forecasting, and understanding customer purchasing patterns. This metric is crucial for financial analysis and reporting.

4. **item_code**
   - **Unique Values**: 27,322 (2012) vs. 23,281 (2013)
   - **Importance**: This represents the diversity of products sold. A high number of unique item codes can indicate a wide product range, which is important for market analysis and product management. It can also help in identifying best-selling items and inventory needs.

5. **warehouse_code**
   - **Unique Values**: 55 (2012) vs. 29 (2013)
   - **Importance**: This indicates the number of warehouses involved. Understanding warehouse distribution is essential for logistics and supply chain management. More warehouses can improve delivery efficiency and customer satisfaction.

6. **contact_method_code**
   - **Unique Values**: 788 (2012) vs. 1,274 (2013)
   - **Importance**: This column reflects the various methods used to contact customers (e.g., email, phone). A diverse range of contact methods can enhance customer engagement strategies. It also helps in analyzing which channels are more effective for communication.

### Columns with Lower Unique Values:
- **company_code**: While the number of unique company codes is fairly low, it could still be important depending on the context (e.g., if there are different product lines or business units).
- **currency**: A limited number of unique currencies indicates a simpler financial context but could be significant for international transactions and pricing strategies.
- **customer_district_code**: The number of districts is relatively low; however, it can provide insights into geographical distribution and regional sales performance.

 The columns with higher unique counts (like `customer_code`, `customer_order_number`, and `item_code`) typically hold more importance for analyzing customer behavior, sales performance, and inventory management. Conversely, columns with fewer unique values still provide context and could be valuable depending on your specific analytical goals.


In [30]:
# Descriptive statistics for the specified columns
stats_2012 = data_2012[columns_to_check].describe(include='all')
stats_2013 = data_2013[columns_to_check].describe(include='all')

print("Descriptive Statistics for 2012 Data:")
print(stats_2012)

print("\nDescriptive Statistics for 2013 Data:")
print(stats_2013)

Descriptive Statistics for 2012 Data:
        company_code customer_code  customer_order_number  invoice_number  \
count   1.037205e+06       1037205           1.037205e+06    1.037205e+06   
unique           NaN          6576                    NaN             NaN   
top              NaN     234750001                    NaN             NaN   
freq             NaN         16705                    NaN             NaN   
mean    2.190146e+02           NaN           1.792116e+06    1.237727e+06   
std     1.715110e+02           NaN           1.453451e+06    1.355065e+06   
min     1.000000e+02           NaN           2.576710e+05    0.000000e+00   
25%     2.050000e+02           NaN           7.354650e+05    6.507800e+05   
50%     2.050000e+02           NaN           1.373569e+06    7.086960e+05   
75%     2.050000e+02           NaN           2.867882e+06    2.195699e+06   
max     9.500000e+02           NaN           9.065931e+06    7.002357e+06   

                             item_cod

The descriptive statistics you’ve provided for the 2012 and 2013 datasets reveal several important insights regarding the chosen columns. Here’s a simplified breakdown of what these statistics indicate:

### General Observations

1. **Data Count**:
   - **2012**: There are **1,037,205** data points in total.
   - **2013**: The total count has decreased to **951,177**. This may suggest a drop in transactions or data entries from one year to the next.

2. **Unique Values**:
   - The number of unique customer codes, orders, and item codes is significant, indicating a diverse customer base and a wide range of products being sold. For example, there are **6,576** unique customer codes in 2012 and **3,915** in 2013.

### Specific Column Insights

1. **company_code**:
   - This column has a consistent range of codes, but the mean value has increased slightly from around **219** in 2012 to about **242** in 2013, which might indicate a slight change in company structure or types of transactions.

2. **customer_code**:
   - In 2012, there were **6,576** unique customer codes, while in 2013, there were only **3,915**. This drop could imply fewer customers or less diverse sales in 2013.

3. **customer_order_number**:
   - The unique order numbers are high, which is typical in retail, but it’s noteworthy that this column has a lot of distinct values, indicating many orders processed over the years.

4. **invoice_number**:
   - Similar to the order numbers, the unique invoice numbers show many transactions, which is important for revenue tracking. The count remains high, but the mean and total are slightly lower in 2013 compared to 2012.

5. **item_code**:
   - There are **27,322** unique items in 2012 and **23,281** in 2013. This decrease might suggest fewer items sold or a reduction in the product range available in 2013.

6. **customer_district_code**:
   - The unique values have slightly increased from **16** in 2012 to **17** in 2013, indicating a stable distribution of sales across districts.

7. **warehouse_code**:
   - The number of unique warehouse codes decreased from **55** to **29**. This could indicate a consolidation of warehouse operations or a change in logistics strategy.

8. **currency**:
   - The currency used is primarily **AUD**, with a slight drop in unique currencies from **6** to **5**, which may indicate less international sales or a shift in currency transactions.

9. **contact_method_code**:
   - The count of unique contact methods increased from **788** to **1,274**, suggesting a wider variety of ways to interact with customers, which could enhance customer service.

10. **line_number**:
    - The line numbers seem to have lower variability and generally remain stable, suggesting consistent transaction entries for items on invoices.

11. **invoice_date**:
    - The mean values for the invoice date column suggest that transactions occurred consistently over both years, with slight differences in transaction timing.

Overall, these statistics can help you understand trends in sales, customer behavior, and operational efficiency between the two years. The decrease in customer codes, item codes, and unique orders from 2012 to 2013 may indicate challenges in maintaining a diverse customer base or a reduction in product offerings. However, the increase in contact methods suggests efforts to improve customer engagement. 

If you need to explore these insights further or want to analyze other aspects, let me know!

Relationships that could be inferred are: 
1. Customer-Related Relationships

customer_code and customer_order_number: The high number of unique customer codes relative to customer order numbers indicates that some customers may place multiple orders. This suggests a potential relationship where returning customers contribute significantly to sales volume.
customer_code and invoice_number: Similar to the previous point, a single customer can have multiple invoices, indicating that customer loyalty or satisfaction may lead to repeated purchases over time.
2. Order and Item Relationships

customer_order_number and item_code: The variety of item codes suggests that customers order multiple different items per order. Analyzing the relationship between these columns could reveal popular item combinations or trends in purchasing behavior.
line_number and item_code: The presence of multiple line numbers for each invoice implies that each invoice can include multiple items. Analyzing line items could help identify which items are frequently ordered together, leading to insights on bundling opportunities or inventory management.
3. Company and Geographic Relationships

company_code and customer_district_code: The relationship between the company codes and customer district codes can indicate regional preferences for certain companies or brands. This can be useful for targeted marketing strategies and stock distribution.
warehouse_code and customer_district_code: With changes in the unique warehouse codes, analyzing this relationship can help understand how geographic distribution affects inventory management and logistics.
4. Temporal Relationships

invoice_date and customer_order_number: Examining how the number of customer orders fluctuates over time (invoice dates) may reveal seasonal trends in purchasing behavior, helping to forecast future sales and manage inventory effectively.
invoice_date and line_number: Analyzing the number of line items over time can help identify trends in purchasing patterns, such as whether customers tend to buy more items during certain periods (e.g., holidays or promotions).
5. Currency and Contact Method Relationships

currency and contact_method_code: Analyzing how the contact method varies with different currencies could reveal insights into customer preferences for payment methods or how the sales process changes based on currency.

In [33]:
# List the specific columns you want to check
specific_columns = [
    'company_code', 
    'customer_code', 
    'customer_order_number', 
    'invoice_number', 
    'item_code', 
    'customer_district_code', 
    'warehouse_code', 
    'currency', 
    'contact_method_code', 
    'line_number', 
    'invoice_date'
]

# Display data types for the specific columns
print(data[specific_columns].dtypes)


company_code                       int64
customer_code                     object
customer_order_number             object
invoice_number                     int64
item_code                         object
customer_district_code             int64
warehouse_code                    object
currency                          object
contact_method_code               object
line_number                        int64
invoice_date              datetime64[ns]
dtype: object


Data Types Breakdown
int64: This data type indicates that the column contains 64-bit integers.

company_code: A numerical identifier for a company.
invoice_number: A numerical identifier for each invoice.
customer_district_code: A numerical identifier for a customer district.
line_number: Typically represents the line item number on an invoice.
object: This data type is typically used for text or mixed data types (strings).


customer_code: A unique identifier for customers, likely a mix of letters and numbers.
customer_order_number: The order number associated with the customer, which might not be purely numeric.
item_code: A code for items being sold, which may include alphanumeric characters.
warehouse_code: The identifier for the warehouse where items are stored.
currency: Represents the currency type (e.g., USD, EUR).
contact_method_code: Codes indicating how the customer was contacted.
datetime64[ns]: This indicates that the column contains date and time information.

invoice_date: This column holds the date and time when the invoice was generated.

Integer columns (int64): Numeric identifiers.
Object columns: Textual or mixed-type identifiers, often alphanumeric.
Datetime column: Holds date and time values.

Implications for Cleaning
Numeric columns: Need to check for any non-numeric values or outliers.
Object columns: Need to look for inconsistencies in format (e.g., different cases in strings), missing values, or unexpected characters.
Datetime column: Have to ensure all dates are in the correct format and check for missing or invalid dates.


## Checking Object Type Columns

In [36]:
# List of object columns to check
object_columns = ['customer_code', 'customer_order_number', 'item_code', 
                  'warehouse_code', 'currency', 'contact_method_code']

# Loop through each object column to check for inconsistencies
for column in object_columns:
    print(f"Checking column: {column}")

    # Check for missing values
    missing_values = data[column].isnull().sum()
    print(f"Missing values: {missing_values}")

    # Convert the column to string to avoid type issues
    data[column] = data[column].astype(str)

    # Check for unique values and their counts
    unique_values = data[column].value_counts()
    print(f"Unique values in '{column}':\n{unique_values}\n")

    # Check for inconsistent cases (e.g., lowercase, uppercase)
    case_counts = data[column].str.lower().value_counts()
    print(f"Case-insensitive counts for '{column}':\n{case_counts}\n")

    # Check for unexpected characters using a regex pattern
    unexpected_chars = data[column][~data[column].str.match(r'^[\w\s-]*$')]
    print(f"Unexpected characters in '{column}':\n{unexpected_chars}\n")



Checking column: customer_code
Missing values: 0
Unique values in 'customer_code':
customer_code
234750001    61844
525200100    11773
359500601     9007
138845501     8935
471000829     8648
             ...  
403400600        1
622000100        1
516200200        1
505480200        1
505470100        1
Name: count, Length: 4487, dtype: int64

Case-insensitive counts for 'customer_code':
customer_code
234750001    61844
525200100    11773
359500601     9007
138845501     8935
471000829     8648
             ...  
403400600        1
622000100        1
516200200        1
505480200        1
505470100        1
Name: count, Length: 4487, dtype: int64

Unexpected characters in 'customer_code':
Series([], Name: customer_code, dtype: object)

Checking column: customer_order_number
Missing values: 0
Unique values in 'customer_order_number':
customer_order_number
263824     301
258450     277
257823     265
264627     264
265611     262
          ... 
2905142      1
2903306      1
2891563      

It implies for cleaning each column:

### 1. **customer_code**
- **Missing Values**: None, so no need to handle missing data.
- **Unique Values**: 4487 unique customer codes, which indicates diversity.
- **Unexpected Characters**: None found, meaning the data is clean.

**Action**: No immediate action needed for this column.

---

### 2. **customer_order_number**
- **Missing Values**: None.
- **Unique Values**: 648780 unique order numbers, indicating a large dataset.
- **Unexpected Characters**: Found a string with a single quote (`'`), which suggests an anomaly.

**Action**: Investigate and clean any unexpected characters. We may want to remove or replace the problematic entry.

---

### 3. **item_code**
- **Missing Values**: None.
- **Unique Values**: 34472 unique item codes.
- **Unexpected Characters**: Several entries contain extra spaces or unusual characters (e.g., `/` and spaces).

**Action**:We can clean the data by trimming whitespace and removing or standardizing unexpected characters.

---

### 4. **warehouse_code**
- **Missing Values**: None.
- **Unique Values**: Numerous unique codes.
- **Unexpected Characters**: None found.

**Action**: No action needed as the data is clean.

---

### 5. **currency**
- **Missing Values**: None.
- **Unique Values**: Multiple currency codes; however, there’s an entry with just a space which might be an issue.
- **Unexpected Characters**: None found.

**Action**: Investigate the entry with just a space and consider removing or replacing it.

---

### 6. **contact_method_code**
- **Missing Values**: None.
- **Unique Values**: 1665 unique contact methods.
- **Unexpected Characters**: Multiple entries contain unusual characters (`&`, `@`, etc.).

**Action**: We can clean the data by standardizing the format for this column and possibly replacing or removing invalid entries.

---

### Data Cleaning Steps:
1. **Trim Whitespace**: For all string columns, trim any leading or trailing whitespace.
2. **Handle Unexpected Characters**: Review and address the unexpected characters in `customer_order_number`, `item_code`, `currency`, and `contact_method_code`.
3. **Standardize Formats**: Ensure that codes (like warehouse codes and currency codes) follow a consistent format (e.g., upper/lowercase).
4. **Duplicate Check**: Given the high number of unique values in some columns, check for duplicates if necessary, especially in `customer_order_number` and `item_code`.

By addressing these issues, we can improve the overall quality and reliability of your dataset for further analysis.

In [39]:
import pandas as pd

# Assuming your DataFrame is named 'df'
# Trim whitespace for all string columns
df = data.apply(lambda x: x.str.strip() if x.dtype == "object" else x)

# If you want to check the result
print(df.head())


   accounting_date  fiscal_year  fiscal_month  calendar_year  calendar_month  \
0         20120509         2012            11           2012               5   
1         20120216         2012             8           2012               2   
2         20120509         2012            11           2012               5   
3         20120518         2012            11           2012               5   
4         20120109         2012             7           2012               1   

   calendar_day  company_code customer_code  customer_district_code  \
0             9           101     411800601                     410   
1            16           101     361000403                     300   
2             9           101     361000403                     300   
3            18           101     565540415                     500   
4             9           101     565540415                     500   

     item_code  ... value_quantity value_price_adjustment currency  \
0  GENIE8WWWBC  ...   

In [40]:
columns_to_trim = ['customer_code', 'customer_order_number', 'item_code', 'warehouse_code', 'currency', 'contact_method_code']
df[columns_to_trim] = df[columns_to_trim].apply(lambda x: x.str.strip())